File: generated-code/javascript/codecraft/TcpReadWrite/model/attack-action.js
const AutoAttack = require.main.require('./model/auto-attack');
/**
 * Attack action
 */
class AttackAction {
    /**
     * If specified, target entity's ID
     */
    target;
    /**
     * If specified, configures auto attacking
     */
    autoAttack;

    constructor(target, autoAttack) {
        this.target = target;
        this.autoAttack = autoAttack;
    }

    /**
     * Read AttackAction from input stream
     */
    static async readFrom(stream) {
        let target;
        if (await stream.readBool()) {
            target = await stream.readInt();
        } else {
            target = null;
        }
        let autoAttack;
        if (await stream.readBool()) {
            autoAttack = await AutoAttack.readFrom(stream);
        } else {
            autoAttack = null;
        }
        return new AttackAction(target, autoAttack);
    }

    /**
     * Write AttackAction to output stream
     */
    async writeTo(stream) {
        let target = this.target;
        if (target === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await stream.writeInt(target);
        }
        let autoAttack = this.autoAttack;
        if (autoAttack === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await autoAttack.writeTo(stream);
        }
    }
}
module.exports = AttackAction

File: generated-code/python/codecraft/TcpReadWrite/model/attack_action.py
from model.auto_attack import AutoAttack

class AttackAction:
    """Attack action"""

    def __init__(self, target, auto_attack):
        self.target = target
        """If specified, target entity's ID"""
        self.auto_attack = auto_attack
        """If specified, configures auto attacking"""

    @staticmethod
    def read_from(stream):
        """Read AttackAction from input stream
        """
        if stream.read_bool():
            target = stream.read_int()
        else:
            target = None
        if stream.read_bool():
            auto_attack = AutoAttack.read_from(stream)
        else:
            auto_attack = None
        return AttackAction(target, auto_attack)
    
    def write_to(self, stream):
        """Write AttackAction to output stream
        """
        if self.target is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_int(self.target)
        if self.auto_attack is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.auto_attack.write_to(stream)
    
    def __repr__(self):
        return "AttackAction(" + \
            repr(self.target) + \
            ", " + \
            repr(self.auto_attack) + \
            ")"